personaelabs/nym
# [READY FOR REVIEW] Add merkle_tree/

Ticket [https://github.com/personaelabs/nym/issues/6](https://github.com/personaelabs/nym/issues/6)

Add a backend service that constructs and saves public key Merkle trees to a database.


## Request For Feedback

- How often should the Merkle tree be updated? In the current implementation, no polling is executed (not using nerman), hence the tree is updated as often as defined in the cron job.
- Are there any noteworthy learnings from running merkle_cron v1 in production? (e.g. common bugs, etc) I feel critical unknown unknowns can easily exist for backend services like this one. 

## How to Test

1. Run commands as described in [README.md](https://github.com/personaelabs/nym/tree/dan/merkle_tree/packages/merkle_tree#readme). The actual Merkle tree construction could take some time.
2. Verify that database tables are filled.

## Notes
- In v1 async-lock is used for concurrency safety, but in this implementation we don’t manage concurrency at the code level. We expect the job to be executed by [Render Cron Jobs](https://render.com/docs/cronjobs), which [guarantees that at most one run is active at any given time](https://render.com/docs/cronjobs#cron-job-concurrency--timeouts). However, if we end up doing polling or using webhooks, we should NOT solely rely on the guarantees of Render.
- Only Safe multisig is supported for now ([code](https://github.com/personaelabs/nym/blob/f39690af65a48dac152b48c40deb2eec748c417a/packages/merkle_tree/src/multisig.ts#L5)).
